File: backend/satisfactory-blueprint/satisfactory-blueprint/src/main/java/com/satisfactory/blueprint/entity/Building.java
package com.satisfactory.blueprint.entity;

import com.satisfactory.blueprint.entity.enums.BuildingType;
import jakarta.persistence.*;
import lombok.Data;

@Entity
@Table(name = "buildings")
@Data
public class Building {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Type of building (e.g. SMELTER, CONSTRUCTOR, etc.).
     * Must be unique.
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, unique = true)
    private BuildingType type;

    /**
     * Sort order index for front-end display (lower values appear first).
     */
    @Column(nullable = false)
    private int sortOrder;

    /**
     * Base power usage in megawatts, editable by the user.
     */
    @Column(nullable = false)
    private double powerUsage;

    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    @JoinColumn(name = "image_id", foreignKey = @ForeignKey(name = "fk_building_image"))
    private Image image;
}

File: frontend/satisfactory-blueprint-designer/src/types/building.ts
// src/types/building.ts

import { type BuildingType } from './enums';
import { type ImageDto } from './image';

/** Matches com.satisfactory.blueprint.dto.BuildingDto */
export interface BuildingDto {
  id: number;
  type: BuildingType;
  sortOrder: number;
  powerUsage: number;
  /** The associated image, if any */
  image?: ImageDto;
}
